File: mahasiswa.php
<?php

abstract class Mahasiswa {
    // Properti terenkapsulasi (protected)
    // Dipetakan pas dari kolom tabel database MySQL sebelumnya
    protected int $id_mahasiswa;
    protected string $nama_mahasiswa; // Opsional: Tambahan nama untuk identitas objek
    protected string $nim;
    protected int $semester;
    protected float $tarifUKTNominal; // Menggunakan float/double untuk tipe DECIMAL di DB

    // Constructor untuk menginisialisasi nilai properti saat objek dibuat (pemetaan dari DB)
    public function __construct(int $id_mahasiswa, string $nama_mahasiswa, string $nim, int $semester, float $tarifUKTNominal) {
        $this->id_mahasiswa = $id_mahasiswa;
        $this->nama_mahasiswa = $nama_mahasiswa;
        $this->nim = $nim;
        $this->semester = $semester;
        $this->tarifUKTNominal = $tarifUKTNominal;
    }

    // --- ABSTRACT METHODS ---
    // Metode abstrak wajib diimplementasikan oleh kelas anak (Mandiri, Bidikmisi, Prestasi)
    
    /**
     * Menghitung total tagihan yang harus dibayar pada semester berjalan
     * (Misal: Mandiri bayar penuh, Bidikmisi Rp0, Prestasi mendapat potongan)
     */
    abstract public function hitungTagihanSemester(): float;

    /**
     * Menampilkan spesifikasi atau detail akademik yang unik berdasarkan jenis pembiayaan
     */
    abstract public function tampilSpesifikasiAkademik(): void;


    // --- GETTER & SETTER (Opsional namun direkomendasikan untuk properti protected) ---
    public function getIdMahasiswa(): int {
        return $this->id_mahasiswa;
    }

    public function getNim(): string {
        return $this->nim;
    }

    public function getSemester(): int {
        return $this->semester;
    }

    public function getTarifUKTNominal(): float {
        return $this->tarifUKTNominal;
    }
}

File: mahasiswa.java
// Abstract Class Mahasiswa
public abstract class Mahasiswa {

    // Atribut terenkapsulasi (protected)
    protected int id_mahasiswa;
    protected String nama_mahasiswa;
    protected String nim;
    protected int semester;
    protected double tarifUKTNominal;

    // Constructor
    public Mahasiswa(int id_mahasiswa, String nama_mahasiswa, String nim,
                     int semester, double tarifUKTNominal) {
        this.id_mahasiswa = id_mahasiswa;
        this.nama_mahasiswa = nama_mahasiswa;
        this.nim = nim;
        this.semester = semester;
        this.tarifUKTNominal = tarifUKTNominal;
    }

    // Getter dan Setter
    public int getId_mahasiswa() {
        return id_mahasiswa;
    }

    public void setId_mahasiswa(int id_mahasiswa) {
        this.id_mahasiswa = id_mahasiswa;
    }

    public String getNama_mahasiswa() {
        return nama_mahasiswa;
    }

    public void setNama_mahasiswa(String nama_mahasiswa) {
        this.nama_mahasiswa = nama_mahasiswa;
    }

    public String getNim() {
        return nim;
    }

    public void setNim(String nim) {
        this.nim = nim;
    }

    public int getSemester() {
        return semester;
    }

    public void setSemester(int semester) {
        this.semester = semester;
    }

    public double getTarifUKTNominal() {
        return tarifUKTNominal;
    }

    public void setTarifUKTNominal(double tarifUKTNominal) {
        this.tarifUKTNominal = tarifUKTNominal;
    }

    // Abstract Method
    public abstract double hitungTagihanSemester();

    public abstract void tampilSepesifikasiAkademik();
}